joblib/parallel.py

```diff
@@ -774,65 +793,153 @@ def print_progress(self):
+            # was very quick and its callback already dispatched all the
+            # remaining jobs.
+            self._iterating = False
+            if self.dispatch_one_batch(iterator):
+                self._iterating = self._original_iterator is not None
+
+            while self.dispatch_one_batch(iterator):
+                pass
+
+            if pre_dispatch == "all" or n_jobs == 1:
+                # The iterable was consumed all at once by the above for loop.
+                # No need to wait for async callbacks to trigger to
+                # consumption.
+                self._iterating = False
 
-                    raise exception
+        except BaseException as exception:
+            try:
+                self._handle_exception(exception)
+            except BaseException:
+                self._terminate_and_reset()
+                raise
+
+    def _next_output(self):
+        # We need to be careful: the buffer list can be filling
+        # up as we empty it and Python list are not thread-safe
+        # up default hence the use of the lock
+        with self._lock:
+            job = self._jobs.pop(0)
+        if getattr(
+                self._backend, 'supports_timeout', False):
+            return job.get(timeout=self.timeout)
+        else:
+            return job.get()
+
+    def _retrieve(self):
+        try:
+            with self._backend.retrieval_context():
+                try:
+                    while self._iterating or len(self._jobs) > 0:
+
+                        if len(self._jobs) == 0:
+                            # Wait for an async callback to dispatch new jobs
+                            time.sleep(0.01)
+                            continue
+
+                        if not self._iterating and len(self._jobs) == 1:
+                            self._pending_outputs.extend(self._next_output())
+                            continue
+
+                        for res in self._next_output():
+                            yield res
+
+                # Note: we catch any BaseException instead of just
+                # Exception instances to also include
+                # KeyboardInterrupt and GeneratorExit
+                except BaseException as exception:
+                    self._handle_exception(exception)
+        finally:
+            pending_outputs = self._pending_outputs
+            self._terminate_and_reset()
+            # Make sure that we get a last message telling us we are done
+            elapsed_time = time.time() - self._start_time
+            self._print('Done %3i out of %3i | elapsed: %s finished',
+                        (self.n_completed_tasks, self.n_completed_tasks,
+                         short_format_time(elapsed_time)))
+            while len(pending_outputs) > 0:
+                yield pending_outputs.pop(0)
+
+    def _retrieve_one_ahead(self):
+        retriever = self._retrieve()
+        try:
+            previous_out = next(retriever)
```

Comment:
With this, the `__call__` function will block until the first task complete.
Isn't there a way to start the computations without starting to iterate on the generator?

I think the issue before was that you did start the computation inside the generator `_retrieve`.
Now that the computation are started in `_start`, you can probably just just return the generator, without having to get the first item.

Reply:
About `_retrieve_one_ahead`, it was indeed unnecessary.